### PR TITLE
Added new method getSkuFromName

### PR DESCRIPTION
**Pull request by arik123**
This method can be used to parse item name, and use it to search

### CODE REVIEW

File: lib/schema.js
Comment (idinium96) [inline lib/schema.js]: Should also do for `Strange Genuine`, `Strange Vintage` (except if "Strange Vintage Tyrolean" but not "Strange Vintage Vintage Tyrolean", and "Strange Vintage Merryweather" but not "Strange Vintage Vintage Merryweather"), and `Strange Unique`

image

File: lib/schema.js
Comment (idinium96) [inline lib/schema.js]: Why is this? 🤔

Comment (arik123) [reply lib/schema.js]: Ctrl + C ... Ctrl + V

File: lib/schema.js
Comment (arik123) [reply lib/schema.js]: Should be fixed for every combination of strange + <quality> are there any other combinations ?

Comment (idinium96) [reply lib/schema.js]: I guess Strange is done.

File: lib/schema.js
Comment (idinium96) [reply lib/schema.js]: 🤔🤔🤔🤔🤔

**Comment by idinium96** [comment]
What about the:
- [Basic ([click here](https://github.com/TF2Autobot/tf2autobot/blob/master/src/lib/extend/item/getSKU.ts#L346-L372)), Specialized, Professional] Killstreak \<name\> Kit;
- [Specialized, Professional] Killstreak \<name\> Kit Fabricator;
- \<name\> Strangifier;
    - got two different types:
        -  General:
            - defindex - 6522
        - Specific:
            - Pomson 6000 Strangifier (defindex - 5661)
            - Pretty Boy's Pocket Pistol Strangifier (defindex - 5721)
            - Phlogistinator Strangifier (defindex - 5722)
            - Cleaner's Carbine Strangifier (defindex - 5723)
            - Private Eye Strangifier (defindex - 5724)
            - Big Chief Strangifier (defindex - 5725)
            - Air Strike Strangifier (defindex - 5753)
            - Classic Strangifier (defindex - 5754)
            - Manmelter Strangifier (defindex - 5755)
            - Vaccinator Strangifier (defindex - 5756)
            - Widowmaker Strangifier (defindex - 5757)
            - Anger Strangifier (defindex - 5758)
            - Apparition's Aspect Strangifier (defindex - 5759)
            - Cow Mangler 5000 Strangifier (defindex - 5783)
            - Third Degree Strangifier (defindex - 5784)
            - Righteous Bison Strangifier (defindex - 5804)
 - \<name\> Strangifier Chemistry Set
     - idk much about this, but I know different defindex:
          - To output Strangifier: `20000`, `20001`, `20005`
- \<name\> Chemistry Set
    - To output Collector's quality item:
        - `20006` (general probably)
        - `20007` (Festive weapons)
        - `20008` (idk)

**Comment by idinium96** [comment]
That's a lot of complex coding. thanks. I forgot one more, painted items.

\<name\> (Paint: \<Paint Name\>)

The `getPaints()` method return an object of `[paintName]: 'p#'`, so maybe can use that to get an array of paint names and then use `getPaintDecimalByName(name)` method to get the value for `item.paint` object.

**Comment by arik123** [comment]
but paints are not in the name normally
if we want to, I could probably just match paint name in search, as there should be no effects with the same name as paints

**Comment by idinium96** [comment]
My bot selling painted cosmetic, so I think if we don't include the bot
might think the other non-painted version.

**Comment by arik123** [comment]
Only chemistry sets are missing now, but I cant find anything about Strangifier Chemistry Set

File: lib/schema.js
Comment (idinium96) [inline lib/schema.js]: `item.paint` is not a string, it's a number type.
so the `paints` object values should be a number (remove the `p`).

**Review by idinium96 — CHANGES_REQUESTED** [review]

**Comment by idinium96** [comment]
> Only chemistry sets are missing now, but I can't find anything about Strangifier Chemistry Set

I think it's okay. I will merge this once you change the paint object values to the number type.

File: lib/schema.js
Comment (idinium96) [inline lib/schema.js]: And also this, might be best to also include extra condition.
`if (name.includes(paint) && name.includes('(Paint: ')) {...}`

File: lib/schema.js
Comment (idinium96) [inline lib/schema.js]: And also, should also replace the "Paint:" and the brackets.
maybe:
`name = name.replace(paint, '').replace('(Paint: ', '').replace(')', '').trim();`

**Review by idinium96 — CHANGES_REQUESTED** [review]

File: lib/schema.js
Comment (idinium96) [reply lib/schema.js]: so it will not detect the actual paint (tool).

Comment (idinium96) [reply lib/schema.js]: Example:
Painted item:
- Vive La France (Paint: Peculiarly Drab Tincture) - sku: 30422;6;p12955537
- Down Under Duster (Paint: An Extraordinary Abundance of Tinge) - sku: 30856;6;p15132390
- Genuine Balloonicorn (Paint: Muskelmannbraun) - sku: 738;1;p10843461
- Strange Arachno-Arsonist (Paint: Color No. 216-190-216) - sku: 31007;11;p14204632

Paint (Tools):
- Color No. 216-190-216 - sku: 5030;6
- Non-Craftable Color No. 216-190-216 - sku: 5030;6;uncraftable
- A Deep Commitment to Purple - sku: 5031;6
- An Extraordinary Abundance of Tinge - sku: 5039;6
- Non-Craftable An Extraordinary Abundance of Tinge - sku: 5039;6;uncraftable

**Comment by arik123** [comment]
Everything should be fine now

File: lib/schema.js
Comment (idinium96) [inline lib/schema.js]: Need to add something here:
```js
    if (name.includes(paintkit) && name.includes(' | ')) {
        name = name.replace(paintkit, '').replace(' | ', '').trim();
        // ...
        }
```

- "Balloonicorn" also exists in the paintkit object.
image
- Unable to get the name properly mainly because of the pipe symbol:
image

Comment (arik123) [reply lib/schema.js]: Could you fix it ? I'm not sure if I can do so

File: lib/schema.js
Comment (idinium96) [inline lib/schema.js]: Need to add some exception here:
- "Haunted Ghosts" (Unusual Effect and a skin name)
- "Haunted Phantasm Jr"
- "Haunted Phantasm"
- "Haunted Metal Scrap"
- "Haunted Hat"
- "Vintage  Tyrolean"
- "Vintage Merryweather"
- "Unusual Cap"

Optional (maybe not need):
- "Brazil Fortress Halloween - Haunted Group Winner"
- "Brazil Fortress Halloween - Haunted Group Participant"
- "Summer 2012 Community Bundle"
- "Halloween 2012 Community Bundle"
- "Halloween 2013 Community Bundle"
- "TF2mixup Community Event 2013 Promo"
- "Halloween 2014 Community Bundle"
- "TF2mixup Community Event 2014 Promo"
- "Tough Break Community Maps Bundle"
- "Robo Community Crate"
- "Robo Community Crate Key"
- "Pile of Robo Community Crate Key Gifts"
- "A Random Robo Community Crate Key Gift"
- "End of the Line Community Crate"
- "Invasion Community Update Pass"
- "Invasion Community Update Coin"
- "Invasion Community Update Key"
- "AU Highlander Community League First Place"
- "AU Highlander Community League Second Place"
- "AU Highlander Community League Third Place"
- "Palaise MGE Doubles Community Clash Champion"
- "Palaise MGE Doubles Community Clash Runner-Up"
- "Palaise MGE Doubles Community Clash Contestant"
- "DreamHack Community Clash 1st Place"
- "DreamHack Community Clash 2nd Place"
- "DreamHack Community Clash 3rd Place"
- "DreamHack Community Clash Participant"
- "DreamHack Community Clash Contributor"
- "End of the Line Community Update Medal"
- "Invasion Community Update Medal"
- ...

Based on all of this, I think the best line of code is:
```js
            const exception = [
                'haunted ghosts',
                'haunted phantasm jr',
                'haunted phantasm',
                'haunted metal scrap',
                'haunted hat',
                'unusual cap'
            ];
            const exception2 = ['vintage tyrolean', 'vintage merryweather'];

            const isNotInException = 
                !exception.some(ex => name.includes(ex)) || !exception2.some(ex2 => name === `${quality} ${ex2}`);

            if (name.startsWith(quality) && isNotInException) {
                name = name.replace(quality, '').trim();
                // ...
            }
```

File: lib/schema.js
Comment (idinium96) [reply lib/schema.js]: Sure thing. I will fix it later.

**Comment by idinium96** [comment]
I am changing the base branch to `getSKUFromName` and I will add you as a maintainer for this repository.

And I will merge this right now.

**Pull request merged**
